qtism/runtime/tests/AssessmentTestSession.php

```diff
@@ -1901,6 +1901,30 @@ protected function nextRouteItem($ignoreBranchings = false, $ignorePreConditions
         }
     }
 
+    /**
+     * Iterate through route items and end test session
+     *
+     * @throws AssessmentItemSessionException
+     * @throws AssessmentTestSessionException
+     * @throws PhpStorageException
+     */
+    public function moveThroughAndEndTestSession()
+    {
+        if ($this->isRunning() === false) {
+            $msg = 'Cannot move to the test ending while the state of the test session is INITIAL or CLOSED.';
+            throw new AssessmentTestSessionException($msg, AssessmentTestSessionException::STATE_VIOLATION);
+        }
+
+        $route = $this->getRoute();
+        while ($route->valid() === true) {
+            $this->nextRouteItem();
+        }
+
+        if ($this->isRunning()) {
```

Comment:
1. Should an exception be thrown in case the condition is `false`?
2. Should the item pointer be reverted to the initial position in case of such an exception?

Reply:
When iteration steps through the route, `nextRouteItem` already calls does the [endTestSession](https://github.com/oat-sa/qti-sdk/blob/legacy/qtism/runtime/tests/AssessmentTestSession.php#L1898). So normally this condition should always evaluate to `false`.